modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.dcache.auth.BearerTokenCredential;
+import org.dcache.auth.attributes.Restriction;
+import org.dcache.gplazma.AuthenticationException;
+import org.dcache.gplazma.plugins.GPlazmaAuthenticationPlugin;
+import org.dcache.gplazma.util.JsonWebToken;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+
+public class TokenExchange implements GPlazmaAuthenticationPlugin {
+
+    private final static Logger LOG = LoggerFactory.getLogger(TokenExchange.class);
+
+    public final static String TOKEN_EXCHANGE_URL = "gplazma.oidc-te.url";
+    public final static String CLIENT_ID = "gplazma.oidc-te.client-id";
+    public final static String CLIENT_SECRET = "gplazma.oidc-te.client-secret";
+    public final static String GRANT_TYPE = "gplazma.oidc-te.grant-type";
+    public final static String SUBJECT_ISSUER = "gplazma.oidc-te.subject-issuer";
+    public final static String SUBJECT_TOKEN_TYPE = "gplazma.oidc-te.subject-token-type";
+    public final static String AUDIENCE = "gplazma.oidc-te.audience";
+
+    private final CloseableHttpClient client;
+
+    private final String tokenExchangeURL;
+    private final String clientID; 
+    private final String clientSecret; 
+    private final String grantType; 
+    private final String subjectIssuer; 
+    private final String subjectTokenType; 
+    private final String audience; 
+
+    public TokenExchange (Properties properties) {
+        this(properties, HttpClients.createDefault());
+
+    }
+
+    @VisibleForTesting
+    TokenExchange (Properties properties, CloseableHttpClient client) {
+
+        tokenExchangeURL = properties.getProperty(TOKEN_EXCHANGE_URL);
+        clientID = properties.getProperty(CLIENT_ID);
+        clientSecret = properties.getProperty(CLIENT_SECRET);
+        grantType = properties.getProperty(GRANT_TYPE);
+        subjectIssuer = properties.getProperty(SUBJECT_ISSUER);
+        subjectTokenType = properties.getProperty(SUBJECT_TOKEN_TYPE);
+        audience = properties.getProperty(AUDIENCE);
+
+        this.client = requireNonNull(client);
+
+    }
+
+    @Override
+    public void authenticate(Set<Object> publicCredentials, Set<Object> privateCredentials,
+        Set<Principal> identifiedPrincipals, Set<Restriction> restrictions)
+        throws AuthenticationException {
+
+        BearerTokenCredential credential = null;
+        for (Object c : privateCredentials) {
+            if (c instanceof BearerTokenCredential) {
+
+                credential = (BearerTokenCredential) c;
+                checkAuthentication(credential == null, "Multiple bearer token credentials");
```

Comment:
the `credential` is never null, as you just have assigned to it a new value. The check must be performed before the assignment (swap the lines).

Reply:
yes, I have swaped the lines. The null check is done before the assignment.